### ISSUE TITLE
Feature request: take pictures

### ISSUE THREAD

**Issue by dannywillems** [codinguser/gnucash-android#503]
Sometimes I would like to take a picture of bills but no way to do it now.

Could be interesting to implement it.

**Comment by bugmenot2** [comment]
Imagine if they could be OCRd too... *whistles*

**Comment by kr4z33** [comment]
To motivate development of this, [I have set up a bounty for this feature](https://freedomsponsors.org/issue/776/feature-request-take-pictures). Open to anyone who can accomplish the task. Feel free to sweeten the offer if you care to.

**Comment by codinguser** [comment]
It would indeed be nice to take a picture of bills. Of course you would still have to manually enter the transaction details in GnuCash, and when you sync with desktop, the images will be ignored. But apart from that, it could work. 

@bugmenot2 OCR is almost completely out of the question.

**Comment by kr4z33** [comment]
If you do keep a long-term feature wishlist for the project somewhere, I suggest adding the concept of GnuCashAndroid OCR on image capturing to it, in case someone decides to make that happen, or if tools become available in the future to do that more easily - it may sound like a long shot now, but it will not hurt to have such long term goals being noted officially as goals by the project. With enough development effort, it should be technically possible to somehow convert some graphic information to text for transaction entry. I have a business card scanner with some proprietary Chinese software that does that with reasonable success.

Regarding the first (hopefully attainable task) of doing image capture and transfer - I suggest consideration of accomplishing the transfer for sync of file attachments from GnuCashAndroid to a Gnucash thus:

1. In GnuCash Android, spawn phone camera to take pictures of things on command, store captured images in a folder "GnuCashAndroidCaptures" alongside the GnuCash Android App, or in a user-designated location on the SD card or whatever.
2. Create a relative path file link "./GnucashAndroidCaptures/file_blah_blah.foo" included for each transaction that has a file association created in the app. (does GnuCash know how to handle relative paths references yet? - I'll assume it does, or will at some point)
3. On data export, create a gnucash Data file, and create also a folder in the same directory called "GnuCashAndroidCaptures" that contains all the files that are linked by the transactions
4. To make the attachments work in GnuCash, the exported data file and it's companion folder would be copied to where the user keeps their Gnucash Data

**Comment by kr4z33** [comment]
Regarding item No. 2 above and some continuity with bugzilla. A Gnucash relative path enhancement request exists.
https://bugzilla.gnome.org/show_bug.cgi?id=767773

**Comment by gabrielsimoes** [comment]
Following up this discussion, as I am planning on getting it going, No.2 has been implemented, but looks like it'll only be available by spring 2017 with release 2.7.

**Comment by codinguser** [comment]
@gabrielsimoes good to hear that you're looking into getting this off the ground.
I would suggest writing out your plan for the feature and how it will impact other parts of the system so that it can be discussed and you get input before starting work. 

Off the top of my head, questions like will this be an Android only feature which is not syncable to desktop. (my guess is yes, because even with No.2 implemented, GnuCash desktop just doesn't support receiving files from elsewhere. Unless we hack it - which is another level of effort). 
- Workflow for image capture and how it will integrate with transactions
- What to do with export to QIF and OFX

And much more...

**Comment by gabrielsimoes** [comment]
Sure.

- I'm planning on allowing the user to input an associated URL (what Gnucash calls "location") or a file from the storage or the camera, which would be copied/saved to a specific folder. 
- A Uri field would keep the relative path to the file inside each transaction. Image capture and deletion would be done inside the transactions form, or something like that. 
- Synchronization would keep the associated string from phone to desktop, but files would only be accessible if the user copied the associated files folder to the desktop and configured the relative path (feature no. 2).
- This way, it might be good to provide a way to export the files to a `.zip` or something.
- I haven't thought much about what to do with QIF and OFX.

**Comment by gabrielsimoes** [comment]
It is rather incomplete, but you can check what I'm trying to do here: https://github.com/gabrielsimoes/gnucash-android/tree/develop

**Comment by storyjesse** [comment]
Great to see this feature has been started! Well done and thank you! :-)
If my programming expertise wasn't limited to bash and .ods sheet formulae I'd give you a hand. I'd be willing to contribute some $$ as an incentive though.

**Comment by ivanskodje** [comment]
@gabrielsimoes Last commit was 9 months ago. Have you abandoned this?

**Comment by gabrielsimoes** [comment]
I'm not working on it now, and probably won't do so until January next year. Thus you can say I abandoned it, sorry.

**Comment by pnemonic78** [comment]
imho:
The image could be stored in the "slot" elements.
The attachment could be an image or document, and encoded as Base64.
The image should be converted to grayscale or black+white to save space.
